url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/PricingPolicy.java

```diff
@@ -0,0 +1,457 @@
+		offpeaks = builder.offpeaks;
+	}
+
+
+	/**
+	 * Gets the billing cycle duration in days.
+	 *
+	 * @return the billing cycle duration in days
+	 */
+	public int getBillingCycle() {
+		return billingCycle;
+	}
+
+	/**
+	 * Gets the fixed charge for every billing cycle.
+	 *
+	 * @return the fixed charge for every billing cycle
+	 */
+	public double getFixedCharge() {
+		return fixedCharge;
+	}
+
+	/**
+	 * Gets the type of the pricing policy.
+	 *
+	 * @return the type of the pricing policy
+	 */
+	public String getType() {
+		return type;
+	}
+
+	/**
+	 * Gets the TOU array.
+	 *
+	 * @return the TOU array
+	 */
+	public double[] getTOUArray() {
+		if(type.equalsIgnoreCase("TOUPricing")) {
+			double[] arr = new double[Constants.MIN_IN_DAY];
+			Iterator<Period> iter = periods.iterator();
+			while(iter.hasNext()) {
+				Period p = iter.next();
+				if(p.getTo().equalsIgnoreCase("00:00")) p.setTo("24:00");
+				String[] fromTokens = p.getFrom().split(":");
+				String[] toTokens = p.getTo().split(":");
+				int from = Integer.parseInt(fromTokens[0]) * 60 + Integer.parseInt(fromTokens[1]);
+				int to = Integer.parseInt(toTokens[0]) * 60 + Integer.parseInt(toTokens[1]);
+				for(int i = from; i < to; i++) {
+					arr[i] = p.getPrice();
+				}
+			}
+			return arr;
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Calculate cost.
+	 *
+	 * @param toEnergy the current energy
+	 * @param fromEnergy the previous energy
+	 * @param toEnergyOffpeak the current energy offpeak
+	 * @param fromEnergyOffpeak the previous energy offpeak
+	 * @param tick the current tick
+	 * @param maxPower the maximum power
+	 * @return the cost
+	 */
+	public double calculateCost(double toEnergy, double fromEnergy,
+			double toEnergyOffpeak, double fromEnergyOffpeak, int tick,
+			double maxPower) {
+		double remainingEnergy = toEnergy - fromEnergy;
+		double cost = 0;
+		switch(type) {
+		case "TOUPricing":
+			int minutesInDay = tick % Constants.MIN_IN_DAY;
+			Iterator<Period> iter = periods.iterator();
+			while(iter.hasNext()) {
+				Period p = iter.next();
+				if(p.getTo().equalsIgnoreCase("00:00")) p.setTo("24:00");
+				String[] fromTokens = p.getFrom().split(":");
+				String[] toTokens = p.getTo().split(":");
+				int from = Integer.parseInt(fromTokens[0]) * 60 + Integer.parseInt(fromTokens[1]);
+				int to = Integer.parseInt(toTokens[0]) * 60 + Integer.parseInt(toTokens[1]);
+				if(minutesInDay >= from && minutesInDay < to) {
+					cost += (toEnergy - fromEnergy) * p.getPrice();
+				}
+			}
+			break;
+		case "ScalarEnergyPricing":
+			cost += fixedCharge;
+			for(int i = levels.size()-1; i >= 0; i--) {
+				double level = levels.get(i).getLevel();
+				double price = levels.get(i).getPrice();
+				if(remainingEnergy < level) {
+					cost += remainingEnergy * price;
+					break;
+				} else if(!(level > 0)) {
+					cost += remainingEnergy * price;
+					break;
+				} else {
+					remainingEnergy -= level;
+					cost += level * price;
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 376:389, 393:406.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.